[MRG] Expose `n_iter_` to BaseLibSVM

jmloyola opened the pull request:
#### Reference Issues/PRs
Fixes #18928

#### What does this implement/fix? Explain your changes.
This PR exposes  `n_iter_` to [BaseLibSVM](https://github.com/scikit-learn/scikit-learn/blob/d4d5f8c7e02cfaced76757fbf38e21c5b28b67b0/sklearn/svm/_base.py#L63):

- Add private attribute `_num_iter` to BaseLibSVM. This will always be a numpy array.
- Add attribute `n_iter_` to BaseLibSVM. When LibSVM optimizes just one model, this is an integer, otherwise, it is a numpy array with the number of iterations for each model.
- Add documentation of `n_iter_` to the classes that inherit from BaseLibSVM (SVR, NuSVR, OneClassSVM, SVC, NuSVC).
- Add `int num_iter` to the class `Solver` in `sklearn/svm/src/libsvm/svm.cpp`. The `Solve` method from this class is responsible for carrying out the optimization. In this method, we save the final number of iterations run in optimization.
- Add `int num_iter` to the struct `decision_function` in `sklearn/svm/src/libsvm/svm.cpp`. Since the svm models do not have direct access to the `Solver` class, we added the attribute `num_iter` to `decision_function`. This struct is returned by the function `svm_train_one` which the train functions call. The function `svm_train_one` calls the Solvers, thus having access to the number of iterations.
- Add `int *num_iter` to `svm_model` and `svm_csr_model` structs in `sklearn/svm/src/libsvm/svm.h`. Since LibSVM can train several models, depending on the model type and the number of classes, here the attribute is an array of integers.
- Fix bug in test `test_libsvm_iris` of `sklearn/svm/tests/test_svm.py`. Since the `fit` function returns a different set of attributes than the ones used by the `predict` function in `sklearn/svm/_libsvm.pyx`, we explicitly selected the correct ones. Note that this didn't fail previously because the `fit_status` variable took the place of the `svm_type` (changing the type of the model).
- Check the attribute `n_iter_` for the classes that inherit from `BaseLibSVM` in `sklearn/utils/estimator_checks.py`. For that I changed the tests `check_non_transformer_estimators_n_iter` and `check_transformer_n_iter`.
- Add changelog
- Test `n_iter_` for all the estimators that inherit from `BaseLibSVM`.

#### Any other comments?

The pre-commit script removed the trailing white spaces in `sklearn/svm/src/libsvm/svm.cpp`.

jmloyola committed:
Expose n_iter_ from libSVM and remove trailing whitespaces (pre-commit)

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Add num_iter to set_model libSVM

jmloyola committed:
Merge upstream/main into add_n_iter_libsvm

jmloyola committed:
Add changelog

jmloyola committed:
Test n_iter_ for BaseLibSVM estimators

jmloyola committed:
Fix failing test

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

ogrisel reviewed (commented):
I started the review but the trailing spaces change make it hard to review this PR. Please undo them to be able to finalize the review.

jjerphan reviewed (commented):
The number of iterations spent in the optimization routine is indeed a useful information to report. :+1: 

Here are some comments and suggestions.

jmloyola committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Revert remotion of trailing white spaces

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Change the type of numpy arrays

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -66,12 +66,54 @@ def test_libsvm_iris():
+        libsvm_probA,
+        libsvm_probB,
+    )
+    pred = _libsvm.predict(iris.data, *lib_svm_model)
```
While coding this, I realized that this will improve code readability for the test, but I will worsen the code for the `predict`, `predict_proba`, `decision_function`, `cross_validation` functions from the Cython code, and all the functions from the `BaseLibSVM` class.

Since we will add a new class as the return type of the fit function, we will have to copy every attribute of `FittedLibSVM` in the `BaseLibSVM` class inside the fit methods (`_dense_fit` and `_sparse_fit`). We can define a method that does this.

Also, for every method of `BaseLibSVM` that depends on the fitted model (`predict`, `predict_proba`, `decision_function`), we will need to create a new instance of `FittedLibSVM` and pass it to the corresponding method. This could also be a new method that creates an instance of `FittedLibSVM` from self.

I wonder if it is not better to leave it like it is now with the unpacked values in the tests. What do you think?

I will try to add the changes so you can see what I mean. Maybe I'm over-complicating things that are simple.

jmloyola replied:
Done, at least for the dense case.
Overall, we added more lines of code with this approach, but I think it is clear.

@jjerphan, Let me know what you think.
Any comments or suggestions are more than welcome.

If it is ok, I will improve the documentation and implement the same extension type for the sparse case.

jjerphan replied:
> While coding this, I realized that this will improve code readability for the test, but I will worsen the code for the predict, predict_proba, decision_function, cross_validation functions from the Cython code, and all the functions from the BaseLibSVM class.

I think you are right.

a44b4a4 add is valid but creates too much changes and complexifies the implementation (I thought it would have been simpler but I was wrong).

I agree with you and think having just the unpacking for this test is better.

In this case a44b4a4 can be reverted and a comment can be added explain why the unpacking and packing are done. For instance (I can't suggest):

```python
    # We unpack the be able to repack some of the
    # return values from libsvm.
    (
        libsvm_support,
        libsvm_support_vectors,
        libsvm_n_class_SV,
        libsvm_sv_coef,
        libsvm_intercept,
        libsvm_probA,
        libsvm_probB,
        # libsvm_fit_status won't be packed bellow.
        libsvm_fit_status,
        libsvm_num_iter,
    ) = _libsvm.fit(iris.data, iris.target.astype(np.float64))

    lib_svm_model = (
        libsvm_support,
        libsvm_support_vectors,
        libsvm_n_class_SV,
        libsvm_sv_coef,
        libsvm_intercept,
        libsvm_num_iter,
        libsvm_probA,
        libsvm_probB,
    )
    pred = _libsvm.predict(iris.data, *lib_svm_model)
```

What do you think?

jmloyola replied:
Done ^^
Btw, congrats @jjerphan on your new role in the project :nerd_face:

jjerphan replied:
Thank you, @jmloyola.

---
"We do what we must because we can."

jmloyola committed:
Add FittedSVMAttributes class to improve code

Apply reviewers suggestions
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Revert "Add FittedSVMAttributes class"

This reverts commit 7a61d14924403b010df63498fd26ef201646dfb4.

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Comment unpacking and repacking of values for test

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan reviewed (commented):
You are nearly done.

I think the only thing left is a test for the correct exposition of `n_iter` when the `n_models=1` as the logic is different in this case.

Also, here are some tiny suggestions.

ogrisel reviewed (commented):
Some more feedback. A general remark would be to try to avoid making the estimator check depend on the estimator names as much as possible.

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola commented on sklearn/svm/src/libsvm/libsvm_helper.c:
```diff
@@ -2,6 +2,13 @@
 #include "_svm_cython_blas_helpers.h"
+
+
+#ifndef max
+    #define max(x, y) (((x) > (y)) ? (x) : (y))
+#endif
```
I agree with you.
I chose to use `max` to be consistent with the other macro definitions in the project. These were defined in some C++ files from the SVM module in the following manner:
```C++
#ifndef max
template <class T> static inline T max(T x,T y) { return (x>y)?x:y; }
#endif
```

What do you think? Do we rename it to `MAX`?

jjerphan replied:
In the case of the snippet you provided, `max` isn't a macro but a template function which is defined conditionally to its absence.

In this PR case, as it is a macro, I would suggest using caps, i.e.:
```suggestion
#ifndef MAX
    #define MAX(x, y) (((x) > (y)) ? (x) : (y))
#endif
```

jjerphan commented on sklearn/svm/src/libsvm/libsvm_sparse_helper.c:
```diff
@@ -3,6 +3,12 @@
 
+
+#ifndef max
+    #define max(x, y) (((x) > (y)) ? (x) : (y))
```
They are needed for correct operator precedence.

For instance, it is better that
```c
max(1+1, 2+2)
```
gets expended to:
```c
(((1+1) > (2+2)) ? (1+1) : (2+2))
```

Though this is a simple example and this might not cause problem for this macro.

jmloyola replied:
As @jjerphan mentioned, they are needed for correct operator precedence.
In case one of `x` or `y` has an operator with less precedence than the relational operators (<, >), the result will not be the expected one.
Though, it is true that in this case, this does not happen and probably will never happen (unless someone calls the macro with something like `max(a&b, c)`).
A simple example that showcases the behavior is:
```C
#include <stdio.h>

#ifndef max
    #define max(x, y) ((x > y) ? x : y)
#endif


int main()
{
    int a, b, c;
    a = 1;
    b = 3;
    c = 2;

    // (((1&3) > (2)) ? (1&3) : (2))
    // x = 1&3
    // y = 2
    // (x > y) ? x : y
    // (1&3 > 2) ? 1&3 : 2
    // (1&1) ? 1&3 : 2
    // 1 ?  1&3 : 2

    printf("1&3 = %d\n", a&b);
    printf("max(1&3, 2) = %d\n", max(a&b, c));

    return 0;
}

```

jmloyola commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3297,7 +3300,9 @@ def check_transformer_n_iter(name, estimator_orig):
         # These return a n_iter per component.
-        if name in CROSS_DECOMPOSITION:
+        if (name in CROSS_DECOMPOSITION) or (
+            name in ["SVC", "NuSVC"] and len(estimator.classes_) > 2
```
You are right. We never get `SVC` or `NuSVC` here. I replicated what I did in `check_non_transformer_estimators_n_iter`. Thanks for catching it.

jmloyola committed:
Apply reviewers suggestions

- Improve `whats_new` message
- Use uppercase for macro name

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jmloyola commented:
The pre-commit script found a trailing white space in [whats_new/v1.1.rst line 59](https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/v1.1.rst). Should I fix it here or leave it to the corresponding [PR](https://github.com/scikit-learn/scikit-learn/pull/21148)?

jmloyola committed:
Apply reviewers suggestions

- Fix check_transformer_n_iter test. It did not need a check for SVC or NuSVC since they are not transformers.
- Simplify the number of iterations assert.
- Use _pairwise_estimator_convert_X if using the precomputed kernel

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented on doc/whats_new/v1.1.rst:
```diff
@@ -131,6 +131,13 @@ Changelog
   Setting a transformer to "passthrough" will pass the features unchanged.
   :pr:`20860` by :user:`Shubhraneel Pal <shubhraneel>`.
 
+:mod:`sklearn.svm`
+...................
+
+- |Enhancement| :class:`svm.OneClassSVM`, :class:`svm.NuSVC`,
+  :class:`svm.NuSVR`, :class:`svm.SVC` and :class:`svm.SVR` now expose
+  `n_iter_`, the number of iterations of the libsvm optimization routine.
+  :pr:`21408` by :user:`Juan Martín Loyola <jmloyola>`.
 :mod:`sklearn.preprocessing`
 ............................
```
This might solve your problem.
```suggestion
:mod:`sklearn.svm`
...................

- |Enhancement| :class:`svm.OneClassSVM`, :class:`svm.NuSVC`,
  :class:`svm.NuSVR`, :class:`svm.SVC` and :class:`svm.SVR` now expose
  `n_iter_`, the number of iterations of the libsvm optimization routine.
  :pr:`21408` by :user:`Juan Martín Loyola <jmloyola>`.

:mod:`sklearn.preprocessing`
............................
```

jmloyola committed:
Update doc/whats_new/v1.1.rst

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan reviewed (approved):
LGTM.

Thank you for this contribution, @jmloyola! To me, you did a good job modifying parts of this old code, asking for clarification and explaining your changes.

jmloyola commented:
Thanks @jjerphan ^^
I think I still need to do another thing in this PR.
Today, after thinking about your comment about checking `n_iter` when the `n_models=1`, I realized the tests do not check the `n_iter` from the `OneClassSVM`
Currently, this is not checked because
```python
getattr(OneClassSVM(), "_estimator_type", None)
# 'outlier_detector'
```
We do not check the number of iterations for this type of estimator.

I will look into it and add a test for this case if needed.

jmloyola committed:
Check n_iter for outlier_detector estimators with max_iter

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola commented:
I added the check `check_non_transformer_estimators_n_iter` to the `'outlier_detector'` estimators.

Now, the `SGDOneClassSVM` and `OneClassSVM` number of iterations are checked.
As far as I understand, these are the only outlier detectors with `n_iter` and `max_iter` attributes.

adrinjalali commented on sklearn/svm/_base.py:
```diff
@@ -266,6 +266,15 @@ def fit(self, X, y, sample_weight=None):
             self.intercept_ *= -1
             self.dual_coef_ = -self.dual_coef_
 
+        # If the number of models optimized by libSVM is one, get the number of
+        # iterations as an integer instead of ndarray.
+        if (
+            self._impl in ["c_svc", "nu_svc"] and len(self.classes_) <= 2
+        ) or self._impl in ["one_class", "epsilon_svr", "nu_svr"]:
```
If this is just to check whether the number of models is 1, why not test for `len(self._num_iter)` instead then? Unless it's the case that sometimes `len(self._num_iter)` is 1, but we still want to expose an ndarray.

jmloyola replied:
You are right. I was trying to be super explicit about this, so we know when (which combination of SVM type and the number of classes) we will have an int and when we will have an array. But I think your suggestion leaves a cleaner code.

adrinjalali commented on sklearn/svm/_classes.py:
```diff
@@ -670,6 +670,15 @@ class SVC(BaseSVC):
         .. versionadded:: 1.0
 
+    n_iter_ : int or ndarray of shape (n_class * (n_class - 1) // 2,)
+        Number of iterations run by the optimization routine to fit the model.
+        If `classes_ <= 2`, only one model is optimized, thus an integer is
+        returned. Otherwise, multiple models are optimized separately, thus
+        having multiple number of iterations. In this case a numpy array is
+        returned with the number of iterations for each model.
```
have we discussed weather this is a good idea? Here we're depending the type of the output to the input, which I'm not sure if it's an entirely good idea. Why not always return an ndarray? It would also simplify the implementation logic.

jmloyola replied:
Yes, I think this should be something to discuss.

I think we have three options:
1. Always return an array (or list). This approach advantage is that the code is simpler and we expose the number of iterations of all the models. The problem is that we make it a little harder for the user to interact with the attribute. For example, when training an `SVR`, the `n_iter_` will always have one element, but the user will have to get the only element of the array to get the integer. This could be confusing since for almost every other estimator with a unique `n_iter_` this is not the case. Another estimator that takes this approach is the LogisticRegression.
2. Depending on the number of models optimized, return an array or an integer. The advantage of this is that the user has direct access to the number of iterations when there is only one model and that we always expose the number of iterations of all the models. The disadvantage is that we have two attributes (one public and another private) that represent the same information, i.e., the number of iterations. Also, here we have a little more code to transform the type of the object depending on the number of models. It should be noted that this approach only adds 5 to 6 lines of code compared to the previous option. Other estimators that take this approach are the `ElasticNet` and `Lars` and the subclasses that inherit from them.
3. Always return an integer. When there is more than one model, the maximum number of iteration across all models is returned. This always returns an integer, but we lose information and we have to add more logic to the codebase (more or less the same number of lines as with the previous approach). Another estimator that takes this approach is `FastICA`.

jmloyola replied:
What do you think @adrinjalali, @ogrisel, @jjerphan?

In my case, the only option I would discard is the third one because we lose information.

adrinjalali reviewed (commented):
Otherwise LGTM.

jmloyola committed:
Simplify the number of models check

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jjerphan commented on sklearn/svm/_base.py:
```diff
@@ -276,9 +276,7 @@ def fit(self, X, y, sample_weight=None):
 
         # If the number of models optimized by libSVM is one, get the number of
         # iterations as an integer instead of ndarray.
```
Could the comment be updated accordingly to this new change? For instance:
```suggestion
        # We number of iterations might be an iterable, but it needs
        # to be returned as an integer.
```

jmloyola committed:
Change n_iter_ to always be ndarray for SVC and NuSVC

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola commented:
After talking to @adrinjalali in the DataUmbrella Post-Sprint, we agreed that a better approach would be to return only one type of object for each sub-class. Thus for `SVC` and `NuSVC`, the `n_iter_` attribute is an ndarray because they can have more than one model optimized. If they optimize one model, an ndarray with one element is returned. While for `SVR`, `NuSVR`, and `OneClassSVM`, the `n_iter_` attribute is an integer since they always optimize one model.

Let me know if you think we still need to change something.

adrinjalali commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -66,12 +66,60 @@ def test_libsvm_iris():
+    ) = _libsvm.fit(iris.data, iris.target.astype(np.float64))
+
+    lib_svm_model = (
+        libsvm_support,
+        libsvm_support_vectors,
+        libsvm_n_class_SV,
+        libsvm_sv_coef,
+        libsvm_intercept,
+        libsvm_n_iter,
+        libsvm_probA,
+        libsvm_probB,
+    )
```
passing kwargs, and here having a `dict` would be nice for readability I think

adrinjalali commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -66,12 +66,60 @@ def test_libsvm_iris():
+        libsvm_n_iter,
+    ) = _libsvm.fit(iris.data, iris.target.astype(np.float64), kernel="linear")
+
+    lib_svm_model = (
+        libsvm_support,
+        libsvm_support_vectors,
+        libsvm_n_class_SV,
+        libsvm_sv_coef,
+        libsvm_intercept,
+        libsvm_n_iter,
+        libsvm_probA,
+        libsvm_probB,
```
same here

adrinjalali commented on sklearn/utils/estimator_checks.py:
```diff
@@ -277,6 +277,7 @@ def _yield_outliers_checks(estimator):
         # test if NotFittedError is raised
         if _safe_tags(estimator, key="requires_fit"):
             yield check_estimators_unfitted
+    yield check_non_transformer_estimators_n_iter
```
interesting that this wasn't here, where was the check before then?

jmloyola replied:
I think the number of iterations from the outlier detectors were not checked before. Otherwise `OneClassSVM` would have failed the test since it has an attribute `max_iter` but it does not have `n_iter`

adrinjalali reviewed (commented):
thanks @jmloyola , overall looks quite good to me.

jmloyola commented on sklearn/svm/_libsvm.pyx:
```diff
@@ -350,7 +361,8 @@ def predict(np.ndarray[np.float64_t, ndim=2, mode='c'] X,
                       support.data, support.shape, sv_coef.strides,
-                      sv_coef.data, intercept.data, nSV.data, probA.data, probB.data)
+                      sv_coef.data, intercept.data, nSV.data, probA.data,
+                      probB.data, n_iter.data)
```
You are right @adrinjalali. I think we can remove the `n_iter` parameter from `set_model` and set `model->n_iter = NULL` inside to avoid problems while freeing the auxiliary models created.

jmloyola committed:
Apply reviewers suggestions

- Remove n_iter from set_model, predict, predict_proba, decision_function, and free_model.
- Update documentation of sub-classes SVC and NuSVC.
- Use a dictionary to load parameters in the test.

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Apply black to sklearn/svm/tests/test_svm.py

adrinjalali commented on sklearn/svm/src/libsvm/libsvm_helper.c:
```diff
@@ -218,6 +228,15 @@ npy_intp get_nr(struct svm_model *model)
+ */
+void copy_n_iter(char *data, struct svm_model *model)
+{
+    const int n_models = MAX(1, model->nr_class * (model->nr_class-1)/2);
```
```suggestion
    const int n_models = MAX(1, model->nr_class * (model->nr_class-1) / 2);
```

I think, or remove spaces around `*`, it's just inconsistent now :D

adrinjalali reviewed (commented):
Thanks, other than the nit I think it's okay now :) 

Would you mind running a benchmark on train and predict, on `main` and on this branch, to check if, or how much speed we loos due to added instructions?

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Apply reviewers suggestions

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

jmloyola commented:
This is the output of running `asv` on the svm benchmark:

```bash
$ pwd
/home/jmloyola/github/scikit-learn/asv_benchmarks

$ git log -1 --pretty=oneline
31df365e9d52da922fea6edf90f54c0dfa887de4 (HEAD -> add_n_iter_libsvm, origin/add_n_iter_libsvm) Apply reviewers suggestions

$ asv continuous --factor 1 -b svm upstream/main HEAD
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31df365e <add_n_iter_libsvm> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl.
· Running 12 total benchmarks (2 commits * 1 environments * 6 benchmarks)
[  0.00%] · For scikit-learn commit 31df365e <add_n_iter_libsvm> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[  8.33%] ··· Setting up svm:14                                                                                                     ok
[  8.33%] ··· svm.SVCBenchmark.peakmem_fit                                                                                          ok
[  8.33%] ··· ========= ======
                kernel        
              --------- ------
                linear   270M 
                 poly    270M 
                 rbf     270M 
               sigmoid   270M 
              ========= ======

[ 16.67%] ··· svm.SVCBenchmark.peakmem_predict                                                                                      ok
[ 16.67%] ··· ========= ======
                kernel        
              --------- ------
                linear   205M 
                 poly    205M 
                 rbf     207M 
               sigmoid   207M 
              ========= ======

[ 25.00%] ··· svm.SVCBenchmark.time_fit                                                                                             ok
[ 25.00%] ··· ========= ===========
                kernel             
              --------- -----------
                linear    957±6ms  
                 poly     958±20ms 
                 rbf     947±100ms 
               sigmoid   1.08±0.2s 
              ========= ===========

[ 33.33%] ··· svm.SVCBenchmark.time_predict                                                                                         ok
[ 33.33%] ··· ========= ===========
                kernel             
              --------- -----------
                linear   1.09±0.2s 
                 poly    964±200ms 
                 rbf     3.92±0.1s 
               sigmoid   1.13±0.2s 
              ========= ===========

[ 41.67%] ··· svm.SVCBenchmark.track_test_score                                                                                     ok
[ 41.67%] ··· ========= ====================
                kernel                      
              --------- --------------------
                linear   0.5899537620849096 
                 poly           0.5         
                 rbf     0.564312736443884  
               sigmoid   0.5710382513661202 
              ========= ====================

[ 50.00%] ··· svm.SVCBenchmark.track_train_score                                                                                    ok
[ 50.00%] ··· ========= ====================
                kernel                      
              --------- --------------------
                linear   0.7373393801965231 
                 poly    0.6089324618736384 
                 rbf     0.7382063936685785 
               sigmoid   0.741296518607443  
              ========= ====================

[ 50.00%] · For scikit-learn commit 793e90b3 <add_n_iter_libsvm~1^2> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 58.33%] ··· Setting up svm:14                                                                                                     ok
[ 58.33%] ··· svm.SVCBenchmark.peakmem_fit                                                                                          ok
[ 58.33%] ··· ========= ======
                kernel        
              --------- ------
                linear   270M 
                 poly    270M 
                 rbf     270M 
               sigmoid   270M 
              ========= ======

[ 66.67%] ··· svm.SVCBenchmark.peakmem_predict                                                                                      ok
[ 66.67%] ··· ========= ======
                kernel        
              --------- ------
                linear   205M 
                 poly    205M 
                 rbf     207M 
               sigmoid   207M 
              ========= ======

[ 75.00%] ··· svm.SVCBenchmark.time_fit                                                                                             ok
[ 75.00%] ··· ========= ===========
                kernel             
              --------- -----------
                linear   1.28±0.3s 
                 poly    1.20±0.2s 
                 rbf     1.26±0.2s 
               sigmoid   1.16±0.2s 
              ========= ===========

[ 83.33%] ··· svm.SVCBenchmark.time_predict                                                                                         ok
[ 83.33%] ··· ========= ===========
                kernel             
              --------- -----------
                linear   1.17±0.2s 
                 poly    1.03±0.2s 
                 rbf     4.34±0.3s 
               sigmoid   1.02±0.2s 
              ========= ===========

[ 91.67%] ··· svm.SVCBenchmark.track_test_score                                                                                     ok
[ 91.67%] ··· ========= ====================
                kernel                      
              --------- --------------------
                linear   0.5899537620849096 
                 poly           0.5         
                 rbf     0.564312736443884  
               sigmoid   0.5710382513661202 
              ========= ====================

[100.00%] ··· svm.SVCBenchmark.track_train_score                                                                                    ok
[100.00%] ··· ========= ====================
                kernel                      
              --------- --------------------
                linear   0.7373393801965231 
                 poly    0.6089324618736384 
                 rbf     0.7382063936685785 
               sigmoid   0.741296518607443  
              ========= ====================

       before           after         ratio
     [793e90b3]       [31df365e]
     <add_n_iter_libsvm~1^2>       <add_n_iter_libsvm>
+            270M             270M     1.00  svm.SVCBenchmark.peakmem_fit('poly')
+            270M             270M     1.00  svm.SVCBenchmark.peakmem_fit('sigmoid')
+            270M             270M     1.00  svm.SVCBenchmark.peakmem_fit('linear')
+            270M             270M     1.00  svm.SVCBenchmark.peakmem_fit('rbf')
-            207M             207M     1.00  svm.SVCBenchmark.peakmem_predict('rbf')
-            207M             207M     1.00  svm.SVCBenchmark.peakmem_predict('sigmoid')
-            205M             205M     1.00  svm.SVCBenchmark.peakmem_predict('poly')
-            205M             205M     1.00  svm.SVCBenchmark.peakmem_predict('linear')
-       1.20±0.2s         958±20ms     0.80  svm.SVCBenchmark.time_fit('poly')
-       1.28±0.3s          957±6ms     0.75  svm.SVCBenchmark.time_fit('linear')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

jmloyola committed:
Revert formatting for unchanged line

jmloyola commented:
I also used IPython magic to test the time and memory for `SVC`.

For both branches I used this code:

```python
from sklearn.svm import SVC
from sklearn.datasets import load_iris
# Load data
X, y = load_iris(return_X_y=True)

%timeit SVC(kernel='linear').fit(X, y)
%memit SVC(kernel='linear').fit(X, y)

model = SVC(kernel='linear').fit(X, y)
%timeit model.predict(X)
%memit model.predict(X)
```

The results are:
|                | upstream:main                                                           | jmloyola:add_n_iter_libsvm                                              |
|----------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------|
| fit time       | 524 µs ± 5.55 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) | 535 µs ± 17 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)   |
| fit memory     | peak memory: 120.34 MiB, increment: 0.79 MiB                            | peak memory: 119.38 MiB, increment: 0.18 MiB                            |
| predict time   | 234 µs ± 2.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) | 234 µs ± 1.19 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) |
| predict memory | peak memory: 120.62 MiB, increment: 0.00 MiB                            | peak memory: 119.60 MiB, increment: 0.00 MiB                            |

jmloyola commented:
To test the sparse case I used:

```python
from sklearn.svm import SVC
from sklearn.datasets import load_iris
from scipy import sparse
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
rng = np.random.RandomState(0)
perm = rng.permutation(iris.target.size)
iris.data = iris.data[perm]
iris.target = iris.target[perm]
iris.data = sparse.csr_matrix(iris.data)

%timeit SVC(kernel='linear').fit(iris.data, iris.target)
%memit SVC(kernel='linear').fit(iris.data, iris.target)

model = SVC(kernel='linear').fit(iris.data, iris.target)
%timeit model.predict(iris.data)
%memit model.predict(iris.data)
```

With this results:

|                | upstream:main                                                            | jmloyola:add_n_iter_libsvm                                               |
|----------------|--------------------------------------------------------------------------|--------------------------------------------------------------------------|
| fit time       | 1.38 ms ± 9.75 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) | 1.39 ms ± 8.32 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) |
| fit memory     | peak memory: 122.01 MiB, increment: 0.00 MiB                             | peak memory: 120.33 MiB, increment: 0.17 MiB                             |
| predict time   | 163 µs ± 601 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  | 170 µs ± 499 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  |
| predict memory | peak memory: 122.01 MiB, increment: 0.00 MiB                             | peak memory: 120.53 MiB, increment: 0.00 MiB                             |

jmloyola commented:
I only tested the `SVC` class in a problem where multiple models are optimized.

Let me know if you think more testing and benchmarks would be helpful.

jjerphan commented:
This does not LGTM, but it does look very good to me.

adrinjalali commented:
For the performance hit, WDYT @jeremiedbb @ogrisel ?

jmloyola commented:
Were you able to look at this @ogrisel, @jeremiedbb? What do you think?

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

adrinjalali commented:
Since it's been a while, would you mind merging the latest `main` to make sure we haven't got any new issues here?

ogrisel commented:
> For the performance hit, WDYT @jeremiedbb @ogrisel ?

What performance hit? Reading the benchmark I see no significant change with `main`. Did I miss something?

ogrisel reviewed (commented):
Sorry for being late to review this. The code looks good.

I think it would be nice to add a couple of SVM specific tests for `n_iter_` in addition to the common test.

For instance checking the shape of `n_iter_` when fitting SVC or NuSVC on the iris dataset: the shape should be `(3,)` (3 * 2 // 2) while fitting on a binary classification problem, (possibly derived from iris) the shape should be `(1,)`.

I was also wondering if we couldn't check some properties about the values them selves, for instance the impact of changing `C` or `gamma` or the dataset on the values of `n_iter_`, for instance for a toy binary classification problem by having a look at: https://en.wikipedia.org/wiki/Sequential_minimal_optimization#Optimization_problem

For instance it seems intuitive that the following toy dataset could converge in a single step whatever the value of C:

```python
>>> SVC(kernel="linear").fit([[-1], [1]], [-1, 1]).n_iter_
array([1], dtype=int32)
```

but I am not 100% sure this is guaranteed.

Alternatively looking at the iris dataset: https://scikit-learn.org/stable/auto_examples/decomposition/plot_pca_vs_lda.html, it seems fair that the 0 vs 1 (setosa vs versicolor) and (0 vs 2) (setosa vs virgina) subproblems are much easier than (1 vs 2) (versicolor vs virginica) and therefore the number of iterations of the last subproblem should be significiantly higher which is what we observe for various different kernels and different values of C as long as C is not to low (regularization is not to high):

```python
>>> from sklearn.datasets import load_iris
>>> from sklearn.svm import SVC
>>> X, y = load_iris(return_X_y=True)
>>> SVC(kernel="linear", C=1).fit(X, y).n_iter_
array([13,  3, 31], dtype=int32)
>>> SVC(kernel="linear", C=100).fit(X, y).n_iter_
array([ 13,   3, 206], dtype=int32)
>>> SVC(kernel="linear", C=0.1).fit(X, y).n_iter_
array([17,  5, 26], dtype=int32)
>>> SVC(kernel="linear", C=0.001).fit(X, y).n_iter_
array([50, 50, 50], dtype=int32)
```

We could therefore write and test such as:

```python
X, y = load_iris(return_X_y=True)
n_iter_iris = SVC(kernel="linear", C=100).fit(X, y).n_iter_

# Looking at the iris dataset:
# https://scikit-learn.org/stable/auto_examples/decomposition/plot_pca_vs_lda.html
# one expects that the 0 vs 1 (setosa vs versicolor) and (0 vs 2) (setosa vs virgina)
# subproblems are much easier than (1 vs 2) (versicolor vs virginica) and therefore
# the number of iterations of the last subproblem should be significiantly higher.
n_iter_0v1, n_iter_0v2, n_iter_1v2 = n_iter_iris
assert n_iter_0v1 < n_iter_1v2
assert n_iter_0v2 < n_iter_1v2
```

Other than the testing coverage that can be improved as suggested above, LGTM.

ogrisel committed:
Merge remote-tracking branch 'origin/main' into add_n_iter_libsvm

ogrisel commented:
I merged main to make sure that everything is in order.

ogrisel commented:
Another test idea: we could check the consistency of the `n_iter_` values for the iris 3-way multiclass problem and the 3 `n_iter_` values obtained by training 3 binary classifiers for the 0v1, 0v2 and 1v2 subproblems constructed manually by fancying indexing the iris dataset manually on y as follows:

```python
mask_0v1 = np.isin(y, [0, 1])
n_iter_0v1 = SVC().fit(X[mask_0v1], y[mask_0v1]).n_iter_
```

and so on.

ogrisel commented:
Also note, the fact that over-regularized models have a number of iterations that match the number of samples per class can probably be explained because in that regime, all samples are support vector and the optimization problem is probably degenerate as all dual coef hit the regularization constraint making the problem trivial to optimize with one-iteration per-sample before reaching the KKT convergence criterion:

```python
>>> C=0.001
>>> SVC(kernel="linear", C=C).fit(X, y).n_iter_
array([50, 50, 50], dtype=int32)
>>> svc.n_support_
array([50, 50, 50], dtype=int32)
>>> np.allclose(np.abs(svc.dual_coef_), C)
```

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Convert n_iter_ to int

jmloyola committed:
Merge 'upstream/main' into add_n_iter_libsvm

jmloyola committed:
Add coverage tests

jmloyola commented:
Thanks for the review @ogrisel.
I added the coverage tests and realized they didn't pass. I had to convert `np.int64` to `int`. Now they do :nerd_face:.

Now, I'll work on the rest of the tests.

jmloyola commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -1059,16 +1105,17 @@ def test_svc_bad_kernel():
 
 
-def test_timeout():
+def test_libsvm_convergence_warnings():
```
I changed the name of the test so that it is consistent with the test for liblinear (`test_linear_svm_convergence_warnings`)

ogrisel reviewed (approved):
LGTM, thank you very much for the follow-up!

ogrisel commented:
Note that https://github.com/scikit-learn/scikit-learn/pull/21408#issuecomment-990874164 could have been turned into a test but this is good enough the way it is (unless you are interested in a new PR with this new test).

jmloyola commented:
Thanks so much, @ogrisel, @jjerphan, and @adrinjalali for the review. :nerd_face: 

I will keep working on a new PR to add more tests. I hadn't had too much time lately.
